psignifit/configuration.py

```diff
@@ -0,0 +1,216 @@
+
+    This class contains a set of valid options and the corresponding sanity
+    checks.
+
+    It raises `PsignifitException` if an invalid option is specified or
+    if a valid option is specified with a value outside of the allowed range.
+
+    Note: attributes and methods starting with `_` are considered private and
+    used internally. Do not change them unlsess you know what you are doing
+
+    Note for the developer: if you want to add a new valid option `foobar`,
+    expand the `Conf.valid_opts` tuple (in alphabetical order) and add any
+    check in a newly defined method `def check_foobar(self, value)`, which
+    raises `PsignifitException` if `value` is outside of the accepted range
+    for `foobar`.
+    """
+    # set of valid options for psignifit. Add new attributes to this tuple
+
+    beta_prior: int = 10
+    CI_method: str = 'percentiles'
+    confP: Tuple[float, float, float] = (.95, .9, .68)
+    dynamic_grid: bool = False
+    estimate_type: str = 'MAP'
+    experiment_type: str = ExperimentType.YES_NO.value
+    experiment_choices: Optional[int] = None
+    fast_optim: bool = False
+    fixed_parameters: Optional[Dict[str, float]] = None
+    grid_set_type: str = 'cumDist'
+    instant_plot: bool = False
+    max_bound_value: float = 1e-05
+    move_bounds: bool = True
+    pool_max_blocks: int  = 25
+    pool_max_gap: float = np.inf
+    pool_max_length: float = np.inf
+    pool_xtol: float = 0
+    priors: Optional[Dict[str, Prior]] = dataclasses.field(default=None, hash=False)
+    sigmoid: str = 'norm'
+    stimulus_range: Optional[Tuple[float, float]] = None
+    thresh_PC: float = 0.5
+    verbose: bool = True
+    width_alpha: float = 0.05
+    width_min: Optional[float] = None
+
+    # parameters, if not specified, will be initialize based on others
+    bounds: Optional[Dict[str, Tuple[float, float]]] = None
+    grid_eval: Optional[int] = None
+    uniform_weight: Optional[float] = None
+
+    # parameters which are always initialized based on other parameters
+    grid_steps: Dict[str, int] = None
+    steps_moving_bounds: Dict[str, int] = None
+
+    def __post_init__(self):
+        self.check_parameters()
+
+    @classmethod
+    def from_dict(cls, config_dict: Dict[str, Any]):
+        config_dict = config_dict.copy()
+        return cls(confP=tuple(config_dict.pop('confP')),
+                   **config_dict)
+
+    def as_dict(self) -> Dict[str, Any]:
+        return dataclasses.asdict(self)
+
+    def check_parameters(self):
+        for field in dataclasses.fields(self):
+            checker_name = 'check_' + field.name
+            if hasattr(self, checker_name):
+                checker = getattr(self, checker_name)
+                checker(getattr(self, field.name))
+
+    # template for an option checking method
+    # def check_foobar(self, value):
+    #    if value > 10:
+    #       raise PsignifitException(f'Foobar must be < 10: {value} given!')
```

Comment:
in this template we need an example what to do when you want to set an attribute (the `object.__setattr__` dance). I am not sure the dance is still needed even if we don't use a frozen dataclass. It was needed before because we were defining a custom `__setattr__` but may not be needed anymore here

Reply:
As I 'unfreezed' the datclassees, we dont have to change anything here.